### PR TITLE
implement StatefulSetStartOrdinal

### LINKED ISSUE #1642: [feature request] StatefulSet StartOrdinal implement
**What would you like to be added**:
cherry-pick StatefulSetStartOrdinal and resolve the conflict with the reserveOrdinals.

**Why is this needed**:
Advanced StatefulSet should cherry-pick it to keep compatible with upstream.

### PR DESCRIPTION
**Pull request by ABNER-1**
### Ⅰ. Describe what this PR does

implement StatefulSetStartOrdinal

### Ⅱ. Does this pull request fix one issue?

fixes #1642

### Ⅲ. Describe how to verify it
implement StatefulSetStartOrdinal

### CODE REVIEW

File: pkg/webhook/statefulset/validating/statefulset_validation.go [hunk 1]
```diff
@@ -324,9 +324,10 @@ func ValidateStatefulSetUpdate(statefulSet, oldStatefulSet *appsv1beta1.Stateful
 	statefulSet.Spec.ReserveOrdinals = oldStatefulSet.Spec.ReserveOrdinals
 	statefulSet.Spec.Lifecycle = oldStatefulSet.Spec.Lifecycle
 	statefulSet.Spec.RevisionHistoryLimit = oldStatefulSet.Spec.RevisionHistoryLimit
+	statefulSet.Spec.Ordinals = oldStatefulSet.Spec.Ordinals
 
 	if !apiequality.Semantic.DeepEqual(statefulSet.Spec, oldStatefulSet.Spec) {
-		allErrs = append(allErrs, field.Forbidden(field.NewPath("spec"), "updates to statefulset spec for fields other than 'replicas', 'template', 'reserveOrdinals', 'lifecycle', 'revisionHistoryLimit', 'persistentVolumeClaimRetentionPolicy', `volumeClaimTemplates` and 'updateStrategy' are forbidden"))
+		allErrs = append(allErrs, field.Forbidden(field.NewPath("spec"), "updates to statefulset spec for fields other than 'replicas', 'ordinals', 'template', 'reserveOrdinals', 'lifecycle', 'revisionHistoryLimit', 'persistentVolumeClaimRetentionPolicy', `volumeClaimTemplates` and 'updateStrategy' are forbidden"))
```
Comment (furykerry) [inline pkg/webhook/statefulset/validating/statefulset_validation.go:330]: shall we validate wether start ordinal and reserveOrdinals is conflicting?

Comment (ABNER-1) [reply pkg/webhook/statefulset/validating/statefulset_validation.go:330]: Currently, I have made some adjustments for compatibility in the function getStatefulSetReplicasRange, and have also added a TODO item. The main issue is that I haven't yet figured out the scenarios where these two features could be used together.

File: pkg/controller/statefulset/stateful_set_control.go [hunk 2]
```diff
@@ -424,26 +427,27 @@ func (ssc *defaultStatefulSetControl) updateStatefulSet(
 			}
 		}
 
-		if ord := getOrdinal(pods[i]); 0 <= ord && ord < replicaCount && !reserveOrdinals.Has(ord) {
+		if ord := getOrdinal(pods[i]); podInOrdinalRange(pods[i], set, replicaCount) && !reserveOrdinals.Has(ord) {
```
Comment (furykerry) [inline pkg/controller/statefulset/stateful_set_control.go:430]: is it better to put !reserveOrdinals.Has(ord) in the podInOrdinalRange?

File: pkg/controller/statefulset/stateful_set_control.go [hunk 3]
```diff
@@ -424,26 +427,27 @@ func (ssc *defaultStatefulSetControl) updateStatefulSet(
...
 			// if the ordinal of the pod is within the range of the current number of replicas and not in reserveOrdinals,
 			// insert it at the indirection of its ordinal
-			replicas[ord] = pods[i]
+			replicas[ord-getStartOrdinal(set)] = pods[i]
 
-		} else if ord >= replicaCount || reserveOrdinals.Has(ord) {
-			// if the ordinal is greater than the number of replicas or in reserveOrdinals,
+		} else if ord >= 0 || reserveOrdinals.Has(ord) {
```
Comment (furykerry) [inline pkg/controller/statefulset/stateful_set_control.go:435]: if ord >= 0, the ord is valid, it seems not necessary to check reserveOrdinals

File: pkg/controller/statefulset/stateful_set_utils.go [hunk 4]
```diff
@@ -165,12 +172,12 @@ func claimOwnerMatchesSetAndPod(claim *v1.PersistentVolumeClaim, set *appsv1beta
 		if hasOwnerRef(claim, set) {
 			return false
 		}
-		podScaledDown := ord >= replicaCount || reserveOrdinals.Has(ord)
+		podScaledDown := !podInOrdinalRange(pod, set, replicaCount) || reserveOrdinals.Has(ord)
```
Comment (furykerry) [inline pkg/controller/statefulset/stateful_set_utils.go:175]: if podInOrdinalRange check also include reserveOrdinals, we can just call podInOrdinalRange

**Review by furykerry — COMMENTED** [review]
comment in L370 also need modification
// slice that will contain all Pods such that 0 <= getOrdinal(pod) < replicaCount and not in reserveOrdinals

File: pkg/controller/statefulset/stateful_set_utils.go [hunk 5]
```diff
@@ -710,17 +723,20 @@ func decreaseAndCheckMaxUnavailable(maxUnavailable *int) bool {
...
 	  - 1
       - 3
+      Spec.Ordinals.Start: 2
 */
-// return replicaCount=6, reserveOrdinals={1, 3}
-
+// result is maxReplicaIdx = 5, reserveOrdinals = {1, 3}
+// replicas[5] stores [replica-2, nil(reserveOrdinal 3), replica-4, replica-5, replica-6]
+// todo: maybe we should remove ineffective reserveOrdinals in webhook, reserveOrdinals = {3}
 func getStatefulSetReplicasRange(set *appsv1beta1.StatefulSet) (int, sets.Int) {
```
Comment (furykerry) [inline pkg/controller/statefulset/stateful_set_utils.go:731]: replicas range is not so well defined if start ordinal is non-zero, maybe we could return start, end and reserved ordinals

File: pkg/controller/statefulset/stateful_set_utils.go [hunk 6]
```diff
@@ -700,6 +719,7 @@ func decreaseAndCheckMaxUnavailable(maxUnavailable *int) bool {
 }
 
 // return parameters is replicaCount and reserveOrdinals, and they are used to support reserveOrdinals scenarios.
+// ReplicasRange => [0, replicaMaxOrdinal - startOrdinal)
```
Comment (furykerry) [inline pkg/controller/statefulset/stateful_set_utils.go:722]: plz revise above two lines accordingly

**Pull request merged**